src/logic/map_objects/tribes/worker_programs/findspace.cc

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (C) 2002-2025 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "logic/map_objects/tribes/worker.h"
+
+#include "logic/map_objects/checkstep.h"
+#include "logic/map_objects/findnode.h"
+#include "logic/map_objects/terrain_affinity.h"
+#include "logic/player.h"
+
+namespace Widelands {
+
+/**
+ * Care about the game.forester_cache_.
+ *
+ * Making the run_findspace routine shorter, by putting one special case into its own function.
+ * This gets called many times each time the forester searches for a place for a sapling.
+ * Since this already contains three nested for-loops, I dedided to cache the values for a
+ * cpu/memory tradeoff (hint: Widelands is pretty heavy on my oldish PC).
+ * Since the implementation details of double could vary between platforms, I decided to
+ * quantize the terrain goodness into int16_t. This lowers the footprint of the caching,
+ * and also makes desyncs caused by different floats horribly unlikely.
+ *
+ * The forester_cache_ is sparse, but then, lookups are fast.
+ *
+ * At the moment of writing, map changing is really infrequent (only in two scenarios)
+ * and even those do not affect this. However, since map changes are possible, this
+ * checks the reliability of the cached value with a small probability (~1%), If a
+ * disparency is found, the entire cache is nuked.
+ *
+ * If somebody in the future makes a scenario, where the land first is barren, and then
+ * spots of eden show up, the foresters will not immediately notice (because of the cache).
+ * They will eventually notice, and since the instance is shared between tribes,
+ * all foresters notice this at the same moment, also in network play. I hope this is okay.
+ *
+ */
+int16_t Worker::findspace_helper_for_forester(const Coords& pos, const Map& map, Game& game) {
```

Comment:
could you please add a TODO here, to check whether this si masking out any Terraforming improvement of the terrain for too long and another one to check whether this is recommended for other terrain affinity immovables as well. (currently only done for tree sapling)